### PR TITLE
add execute mod for interpret command

### PR DESCRIPTION
**Pull request by ikaven1024**
Signed-off-by: yingjinhui <yingjinhui@didiglobal.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
add skeleton of interpreter command.

**Which issue(s) this PR fixes**:
part of #2371

**Special notes for your reviewer**:

### Testing

Prepare a `get-replica.yml`

<details>

```yaml
apiVersion: config.karmada.io/v1alpha1
kind: ResourceInterpreterCustomization
metadata:
  name: get-replica
spec:
  target:
    apiVersion: apps/v1
    kind: Deployment
  customizations:
    retention:
    replicaResource:
      luaScript: >
        function GetReplicas(obj)
          replica = obj.spec.replicas + 1
          requirement = {}
          return replica, requirement
        end
```
</details>

and `deploy-nginx.yml`

<details>

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
spec:
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx:alpine
        name: nginx
```
</details>

```bash
karmadactl interpret -f get-replica.yml --observed-file deploy-nginx.yml --operation=replicaResource
```

Output:
```
# replicaResource RESULTS:
---
# REPLICA
4
---
# REQUIRES
nodeclaim: null
resourcerequest: {}
```

**Does this PR introduce a user-facing change?**:
```release-note
`karmadactl`: add execute mod for interpret command.
```

### CODE REVIEW

**Comment by XiShanYongYe-Chang** [comment]
Hi @ikaven1024 as #2750 has been merged, dose this need to rebase the master branch?

**Comment by ikaven1024** [comment]
@RainbowMango @XiShanYongYe-Chang  this PR is ready for reviewing.

**Comment by XiShanYongYe-Chang** [comment]
How about providing some test demos for each action in the library, otherwise it's too difficult for users to construct tests, such as `aggregatedStatus`.

**Comment by XiShanYongYe-Chang** [comment]
> How about providing some test demos for each action in the library, otherwise it's too difficult for users to construct tests, such as `aggregatedStatus`.

@RainbowMango How do you think?

File: pkg/karmadactl/interpret/interpret.go [hunk 1]
```diff
@@ -141,6 +157,56 @@ func (o *Options) Run() error {
...
+		return nil, nil
+	}
+
+	objs, err := o.StatusResult.Objects()
+	if err != nil {
+		return nil, err
+	}
+	items := make([]workv1alpha2.AggregatedStatusItem, len(objs))
+	for i, obj := range objs {
+		items[i] = *(obj.(*workv1alpha2.AggregatedStatusItem))
```
Comment (XiShanYongYe-Chang) [inline pkg/karmadactl/interpret/interpret.go:188]: Is it possible that the user is not providing `AggregatedStatusItem`?

File: pkg/karmadactl/interpret/execute.go [hunk 2]
```diff
@@ -2,15 +2,123 @@ package interpret
...
+
+	for _, r := range o.Rules {
+		ruleName := strings.ToLower(r.Name())
+		if ruleName == strings.ToLower(o.Operation) {
+			result := r.Run(interpreter, args)
+			printExecuteResult(o.Out, o.ErrOut, r.Name(), result)
+			return nil
+		}
+	}
+	return fmt.Errorf("operation %s is not supported. Use one of: %s", o.Operation, strings.Join(o.Rules.Names(), ", "))
```
Comment (XiShanYongYe-Chang) [inline pkg/karmadactl/interpret/execute.go:97]: Can the operation be validated in the `Validate()` function in advance?

**Comment by RainbowMango** [comment]
> How about providing some test demos for each action in the library, 

What kind of demo we are taking about? And, to which library?

**Comment by XiShanYongYe-Chang** [comment]
> What kind of demo we are taking about? 

Demo for `karmadactl interpreter`, specifically, the YAML file during execution.

> And, to which library?

Current library, such as `examples`.

File: pkg/karmadactl/util/genericresource/builder.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+// Builder provides convenience functions for taking arguments and parameters
+// from the command line and converting them to a list of resources to iterate
+// over using the Visitor interface.
+type Builder struct {
+	errs       []error
+	paths      []Visitor
+	stdinInUse bool
+	mapper     *mapper
+	schema     resource.ContentValidator
```
[the remark this answers is not in the document]
Comment (ikaven1024) [reply pkg/karmadactl/util/genericresource/builder.go:30]: Yes, it's reserved to validate the editing file.

File: pkg/karmadactl/interpret/execute.go [hunk 4]
```diff
@@ -1,16 +1,157 @@
...
+			RequireObject(true).
+			Local().
+			Do()
+		errs = append(errs, o.ObservedResult.Err())
+	}
+
+	if len(o.StatusFile) > 0 {
+		o.StatusResult = genericresource.NewBuilder().
+			Constructor(func() interface{} { return &workv1alpha2.AggregatedStatusItem{} }).
+			Filename(true, o.StatusFile).
```
[the remark this answers is not in the document]
Comment (ikaven1024) [reply pkg/karmadactl/interpret/execute.go:44]: Yes, you are right. We only accept one status file.

File: pkg/karmadactl/util/genericresource/visitor.go [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+}
+
+func ignoreFile(path string, extensions []string) bool {
+	if len(extensions) == 0 {
+		return false
+	}
+	ext := filepath.Ext(path)
+	for _, s := range extensions {
+		if s == ext {
+			return false
```
[the remark this answers is not in the document]
Comment (ikaven1024) [reply pkg/karmadactl/util/genericresource/visitor.go:74]: Extension matched, shall NOT be ignored. So here returns false.

File: pkg/karmadactl/util/genericresource/result.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import (
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+// Result contains helper methods for dealing with the outcome of a Builder.
+type Result struct {
+	err     error
+	visitor Visitor
+
+	sources []Visitor
```
[the remark this answers is not in the document]
Comment (ikaven1024) [reply pkg/karmadactl/util/genericresource/result.go:12]: deleted

File: pkg/karmadactl/util/genericresource/visitor.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (XiShanYongYe-Chang) [reply pkg/karmadactl/util/genericresource/visitor.go:74]: okay, thanks~

File: pkg/resourceinterpreter/configurableinterpreter/configmanager/manager.go [hunk 7]
```diff
@@ -63,17 +64,21 @@ func (configManager *interpreterConfigManager) HasSynced() bool {
...
 	}
 	manager.configuration.Store(make(map[schema.GroupVersionKind]CustomAccessor))
 	manager.initialSynced.Store(false)
-	configHandlers := fedinformer.NewHandlerOnEvents(
-		func(_ interface{}) { manager.updateConfiguration() },
-		func(_, _ interface{}) { manager.updateConfiguration() },
-		func(_ interface{}) { manager.updateConfiguration() })
-	inform.ForResource(resourceInterpreterCustomizationsGVR, configHandlers)
+
+	if inform != nil {
```
Comment (RainbowMango) [inline pkg/resourceinterpreter/configurableinterpreter/configmanager/manager.go:73]: When will the `inform` be nil?

By the way, `inform` should be `informer`, it's a typo brought by the previous patch.

File: pkg/karmadactl/interpret/execute.go [hunk 8]
```diff
@@ -1,16 +1,157 @@
...
 	"k8s.io/kubectl/pkg/cmd/util"
+
+	workv1alpha2 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha2"
+	"github.com/karmada-io/karmada/pkg/karmadactl/util/genericresource"
+	"github.com/karmada-io/karmada/pkg/resourceinterpreter/configurableinterpreter"
 )
 
-func (o *Options) completeExecute(_ util.Factory, _ *cobra.Command, _ []string) []error {
-	return nil
+func (o *Options) completeExecute(f util.Factory, _ *cobra.Command, _ []string) []error {
```
Comment (RainbowMango) [inline pkg/karmadactl/interpret/execute.go:19]: Why keep `corbra.Command` and `[]string` in the method signature if we don't use them?

**Comment by XiShanYongYe-Chang** [comment]
The function test is okay.

yaml preparation:

<details>
<summary>observed-deploy-nginx.yaml</summary>

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  replicas: 3
  paused: true
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx
        name: nginx
status:
  availableReplicas: 2
  observedGeneration: 1
  readyReplicas: 2
  replicas: 2
  updatedReplicas: 2
```
</details>

<details>
<summary>desired-deploy-nginx.yaml</summary>

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  replicas: 3
  paused: false
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx
        name: nginx
      serviceAccountName: test-sa
```
</details>

<details>
<summary>config.yaml</summary>

```yaml
apiVersion: config.karmada.io/v1alpha1
kind: ResourceInterpreterCustomization
metadata:
  name: get-replica
spec:
  target:
    apiVersion: apps/v1
    kind: Deployment
  customizations:
    retention:
    replicaResource:
      luaScript: >
        function GetReplicas(obj)
          replica = obj.spec.replicas
          requirement = {}
          return replica, requirement
        end
    replicaRevision:
      luaScript: >
        function ReviseReplica(obj, desiredReplica)
          obj.spec.replicas = desiredReplica
          return obj
        end
    retention:
      luaScript: >
        function Retain(desiredObj, observedObj)
          desiredObj.spec.paused = observedObj.spec.paused
          return desiredObj
        end
    statusAggregation:
      luaScript: >
        function AggregateStatus(desiredObj, statusItems)
          if statusItems == nil then
            return desiredObj
          end
          if desiredObj.status == nil then
            desiredObj.status = {}
          end
          replicas = 0
          for i = 1, #statusItems do
            if statusItems[i].status ~= nil and statusItems[i].status.replicas ~= nil  then
              replicas = replicas + statusItems[i].status.replicas
            end
          end
          desiredObj.status.replicas = replicas
          return desiredObj
        end
    statusReflection:
      luaScript: >
        function ReflectStatus (observedObj)
          if observedObj.status == nil then
            return nil
          end
        return observedObj.status
        end
    healthInterpretation:
      luaScript: >
        function InterpretHealth(observedObj)
          return observedObj.status.readyReplicas == observedObj.spec.replicas
        end
    dependencyInterpretation:
      luaScript: >
        function GetDependencies(desiredObj)
          dependentSas = {}
          refs = {}
          if desiredObj.spec.template.spec.serviceAccountName ~= '' and desiredObj.spec.template.spec.serviceAccountName ~= 'default' then
            dependentSas[desiredObj.spec.template.spec.serviceAccountName] = true
          end
          local idx = 1
          for key, value in pairs(dependentSas) do
            dependObj = {}
            dependObj.apiVersion = 'v1'
            dependObj.kind = 'ServiceAccount'
            dependObj.name = key
            dependObj.namespace = desiredObj.metadata.namespace
            refs[idx] = dependObj
            idx = idx + 1
          end
          return refs
        end
```
</details>

<details>
<summary>status-file.yaml</summary>

```yaml
applied: true
clusterName: member1
health: Healthy
status:
  availableReplicas: 1
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1
---
applied: true
clusterName: member2
health: Healthy
status:
  availableReplicas: 1
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1
```
</details>

test command:

```
# karmadactl interpret -f config.yaml --observed-file observed-deploy-nginx.yaml --operation=reviseReplica --desired-replica 5
---
# [1/1] revised:
apiVersion: apps/v1
kind: Deployment
metadata:
    labels:
        app: nginx
    name: nginx
spec:
    paused: true
    replicas: 5
    selector:
        matchLabels:
            app: nginx
    template:
        metadata:
            labels:
                app: nginx
        spec:
            containers:
                - image: nginx
                  name: nginx
status:
    availableReplicas: 2
    observedGeneration: 1
    readyReplicas: 2
    replicas: 2
    updatedReplicas: 2
```

```
# karmadactl interpret -f config.yaml --observed-file observed-deploy-nginx.yaml --operation=InterpretReplica
---
# [1/2] replica:
3
---
# [2/2] requires:
{}
```

```
# karmadactl interpret -f config.yaml --desired-file desired-deploy-nginx.yaml --observed-file observed-deploy-nginx.yaml --operation retain
---
# [1/1] retained:
apiVersion: apps/v1
kind: Deployment
metadata:
    labels:
        app: nginx
    name: nginx
spec:
    paused: true
    replicas: 3
    selector:
        matchLabels:
            app: nginx
    template:
        metadata:
            labels:
                app: nginx
        spec:
            containers:
                - image: nginx
                  name: nginx
            serviceAccountName: test-sa
```

```
# karmadactl interpret -f config.yaml --observed-file observed-deploy-nginx.yaml --operation interpretStatus
---
# [1/1] status:
availableReplicas: 2
observedGeneration: 1
readyReplicas: 2
replicas: 2
updatedReplicas: 2
```

```
# karmadactl interpret -f config.yaml --observed-file observed-deploy-nginx.yaml --operation InterpretHealth
---
# [1/1] healthy:
false
```

```
# karmadactl interpret -f config.yaml --desired-file desired-deploy-nginx.yaml --operation interpretDependency
---
# [1/1] dependencies:
- apiVersion: v1
  kind: ServiceAccount
  name: test-sa
```

```
# karmadactl interpret -f config.yaml --desired-file desired-deploy-nginx.yaml --operation aggregateStatus --status-file status-file.yaml
---
# [1/1] aggregatedStatus:
apiVersion: apps/v1
kind: Deployment
metadata:
    labels:
        app: nginx
    name: nginx
spec:
    paused: false
    replicas: 3
    selector:
        matchLabels:
            app: nginx
    template:
        metadata:
            labels:
                app: nginx
        spec:
            containers:
                - image: nginx
                  name: nginx
            serviceAccountName: test-sa
status:
    replicas: 2
```

File: pkg/karmadactl/interpret/interpret.go [hunk 9]
```diff
@@ -32,18 +36,30 @@ var (
 	interpretExample = templates.Examples(`
         # Check the customizations in file
         %[1]s interpret -f customization.json --check
+
 		# Execute the retention rule for
```
[the remark this answers is not in the document]
Comment (ikaven1024) [reply pkg/karmadactl/interpret/interpret.go]: removed

File: pkg/karmadactl/interpret/interpret.go [hunk 10]
```diff
@@ -32,18 +36,30 @@ var (
...
+
 		# Execute the dependencyInterpretation rule 
         %[1]s interpret -f customization.yml --operation interpretDependency --observed-file observed.yml
+
 		# Execute the statusAggregation rule
-        %[1]s interpret -f customization.yml --operation aggregateStatus --status-file status1.yml --status-file status2.yml
+        %[1]s interpret -f customization.yml --operation aggregateStatus --observed-file observed.yml --status-file status.yml
+
+		# Fetch arguments from url and stdin
+        %[1]s interpret -f customization.yml --operation aggregateStatus --observed-file https://example.com/observed.yml --status-file -
```
[the remark this answers is not in the document]
Comment (ikaven1024) [reply pkg/karmadactl/interpret/interpret.go]: updated

File: pkg/karmadactl/interpret/interpret.go [hunk 11]
```diff
@@ -81,7 +97,7 @@ func NewCmdInterpret(f util.Factory, parentCommand string, streams genericcliopt
 	flags.BoolVar(&o.Check, "check", false, "Validates the given ResourceInterpreterCustomization configuration(s)")
 	flags.StringVar(&o.DesiredFile, "desired-file", o.DesiredFile, "Filename, directory, or URL to files identifying the resource to use as desiredObj argument in rule script.")
 	flags.StringVar(&o.ObservedFile, "observed-file", o.ObservedFile, "Filename, directory, or URL to files identifying the resource to use as observedObj argument in rule script.")
-	flags.StringSliceVar(&o.StatusFile, "status-file", o.StatusFile, "Filename, directory, or URL to files identifying the resource to use as statusItems argument in rule script.")
+	flags.StringVar(&o.StatusFile, "status-file", o.StatusFile, "Filename, directory, or URL to files identifying the resource to use as statusItems argument in rule script.")
```
[the remark this answers is not in the document]
Comment (ikaven1024) [reply pkg/karmadactl/interpret/interpret.go:100]: Yes, This command is working:
```
➜ curl http://127.0.0.1/status.yml

clusterName: cluster1
status:
  readyReplicas: 5
applied: true
---
clusterName: cluster2
status:
  readyReplicas: 5
applied: true

➜ karmadactl interpret -f customization.yml --operation aggregateStatus --observed-file observed.yml --status-file http://127.0.0.1/status.yml
```

It's implemented in https://github.com/karmada-io/karmada/blob/47cb8a683eda20c5cd6cd0e7db57f0280667f7f9/pkg/karmadactl/util/genericresource/visitor.go#L238

File: pkg/karmadactl/interpret/interpret.go [hunk 12]
```diff
@@ -128,6 +147,12 @@ func (o *Options) Complete(f util.Factory, cmd *cobra.Command, args []string) er
 
 // Validate checks the EditOptions to see if there is sufficient information to run the command.
```
[the remark this answers is not in the document]
Comment (ikaven1024) [reply pkg/karmadactl/interpret/interpret.go]: updated

File: pkg/resourceinterpreter/configurableinterpreter/configmanager/manager.go (hunk 7 above)
Comment (ikaven1024) [reply pkg/resourceinterpreter/configurableinterpreter/configmanager/manager.go]: > When will the `inform` be nil?
In executing mode, config is not synced from informer, it's loaded by
https://github.com/karmada-io/karmada/blob/47cb8a683eda20c5cd6cd0e7db57f0280667f7f9/pkg/karmadactl/interpret/execute.go#L82

So informer is unused in this mode.



> By the way, `inform` should be `informer`, it's a typo brought by the previous patch.

updated

File: pkg/karmadactl/interpret/execute.go (hunk 8 above)
Comment (ikaven1024) [reply pkg/karmadactl/interpret/execute.go]: removed

File: pkg/karmadactl/interpret/interpret.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/karmadactl/interpret/interpret.go]: I still can't understand when will get the observed file from a remote server.
Can you give a example?

[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/karmadactl/interpret/interpret.go]: What's the use case? I mean.

File: pkg/resourceinterpreter/configurableinterpreter/configmanager/manager.go (hunk 7 above)
Comment (RainbowMango) [reply pkg/resourceinterpreter/configurableinterpreter/configmanager/manager.go]: Have you met a case that the `informer` is `nil`?

I'm still confused about why to add the `if inform != nil`.

File: pkg/karmadactl/interpret/interpret.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ikaven1024) [reply pkg/karmadactl/interpret/interpret.go]: This flag is just like `-f` in kubectl. 

```
➜  kubectl create -h

    -f, --filename=[]:
	Filename, directory, or URL to files to use to create the resource
```

I think some guys put their config files on a http server, rather local.

File: pkg/resourceinterpreter/configurableinterpreter/configmanager/manager.go (hunk 7 above)
Comment (ikaven1024) [reply pkg/resourceinterpreter/configurableinterpreter/configmanager/manager.go]: In interpret command, we don't use informer, so it's nil.

Comment (XiShanYongYe-Chang) [reply pkg/resourceinterpreter/configurableinterpreter/configmanager/manager.go]: Maybe we can add some comments to explain the situation where informer is nil.

Comment (ikaven1024) [reply pkg/resourceinterpreter/configurableinterpreter/configmanager/manager.go]: > Maybe we can add some comments to explain the situation where informer is nil.

Updated.

**Comment by XiShanYongYe-Chang** [comment]
Thanks!
/lgtm

File: pkg/karmadactl/interpret/interpret.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/karmadactl/interpret/interpret.go]: ok, will see.

**Review by RainbowMango — APPROVED** [review]
/approve
Thanks.

**Pull request merged**